File: src/com/moneydance/modules/features/formula/split/FormulaSplitTxn.java
package com.moneydance.modules.features.formula.split;

import com.infinitekind.moneydance.model.SplitTxn;
import com.moneydance.modules.features.formula.MDApi;
import lombok.Data;
import org.apache.commons.lang3.StringUtils;

import java.math.BigInteger;

@Data
public class FormulaSplitTxn {
    private int splitIndex;
    private SplitTxn txn;
    private FormulaResolver resolver;
    private String A;
    private String B;
    private String C;

    public FormulaSplitTxn(int splitIndex, SplitTxn txn, FormulaResolver resolver) {
        this.splitIndex = splitIndex;
        this.txn = txn;
        this.resolver = resolver;

        A = txn.getParameter("formula_a");
        B = txn.getParameter("formula_b");
        C = txn.getParameter("formula_c");
    }

    public void syncSettings() {
        long value = getAmount();
        txn.setAmount(value, value);
        txn.setParameter("formula_a", A);
        txn.setParameter("formula_b", B);
        txn.setParameter("formula_c", C);
    }

    private String addParentheses(String formula) {
        if (formula.contains("+") || formula.contains("-")) {
            return "(" + formula + ")";
        }
        return formula;
    }

    private String formula() {
        String add = StringUtils.isNoneBlank(C) ? ("+" + C) : "";
        if (StringUtils.isNoneBlank(A)  && StringUtils.isNoneBlank(B)) {
            return addParentheses(A) + "*" + addParentheses(B) + add;
        } else if (StringUtils.isNoneBlank(A)) {
            return A + add;
        } else if (StringUtils.isNoneBlank(B)) {
            return B + add;
        } else if (StringUtils.isNoneBlank(C)) {
            return C;
        } else {
            return null;
        }
    }

    public static Long toCents(Object value) {
        if (value instanceof Double) {
            return Math.round(100 * (Double) value);
        } else if (value instanceof Float) {
            return (long) Math.round(100 * (Float) value);
        } else if (value instanceof Long) {
            return 100 * (Long)value;
        } else if (value instanceof Integer) {
            return (long) (100 * (Integer) value);
        } else if (value instanceof BigInteger) {
            return 100 * ((BigInteger) value).longValue();
        } else {
            return null;
        }
    }

    public long getAmount() {
        Long value = toCents(resolver.getValue(splitIndex));
        return value != null ? value : txn.getValue();
    }

    public String getDescription() {
        String description = txn.getDescription();
        String formula = formula();
        if (formula != null) {
            description += " (" + formula + ")";
        }
        return description;
    }

    public String formatValue(Object valueOrError, boolean showNegativeOnly) {
        if (valueOrError != null && !(valueOrError instanceof Number)) {
            return valueOrError.toString();
        }

        Long value = toCents(valueOrError);

        long defaultValue = txn.getValue();
        if (value == null) {
            value = defaultValue;
        }

        if (showNegativeOnly && value < 0) {
            // render values as positive
            value *= -1;
            defaultValue *= -1;
        } else if (showNegativeOnly || value < 0) {
            return "";
        }

        if (value != defaultValue) {
            return String.format("%s (was %s)",
                    MDApi.formatCurrency(value),
                    MDApi.formatCurrency(defaultValue));
        } else {
            return MDApi.formatCurrency(value);
        }
    }

    public String toString() {
        return txn.getDescription();
    }

    public String getCellSource(char col) {
        switch (col) {
            case 'A': return A;
            case 'B': return B;
            case 'C': return C;
            case 'V': return formula();
            default: return null;
        }
    }
}

File: formula/src/main/kotlin/com/moneydance/modules/features/formula/split/FormulaSplitTxn.kt
package com.moneydance.modules.features.formula.split

import com.moneydance.modules.features.MDApi.Companion.formatCurrency
import com.infinitekind.moneydance.model.SplitTxn
import org.apache.commons.lang3.StringUtils
import java.math.BigInteger
import kotlin.math.roundToLong

class FormulaSplitTxn(private val splitIndex: Int, val txn: SplitTxn, private val resolver: FormulaResolver) {
    var a: String? = txn.getParameter("formula_a")
    var b: String? = txn.getParameter("formula_b")
    var c: String? = txn.getParameter("formula_c")

    fun syncSettings() {
        val value = amount
        txn.setAmount(value, value)
        txn.setParameter("formula_a", a)
        txn.setParameter("formula_b", b)
        txn.setParameter("formula_c", c)
    }

    private fun addParentheses(formula: String): String {
        return if (formula.contains("+") || formula.contains("-")) {
            "($formula)"
        } else formula
    }

    private fun formula(): String? {
        val add = if (StringUtils.isNoneBlank(c)) "+$c" else ""
        return if (StringUtils.isNoneBlank(a) && StringUtils.isNoneBlank(b)) {
            addParentheses(a!!) + "*" + addParentheses(b!!) + add
        } else if (StringUtils.isNoneBlank(a)) {
            a + add
        } else if (StringUtils.isNoneBlank(b)) {
            b + add
        } else if (StringUtils.isNoneBlank(c)) {
            c
        } else {
            null
        }
    }

    val amount: Long
        get() {
            val value = toCents(resolver.getValue(splitIndex))
            return value ?: txn.value
        }
    val description: String
        get() {
            var description = txn.description
            val formula = formula()
            if (formula != null) {
                description += " ($formula)"
            }
            return description
        }

    fun formatValue(valueOrError: Any?, showNegativeOnly: Boolean): String {
        if (valueOrError != null && valueOrError !is Number) {
            return valueOrError.toString()
        }
        var value = toCents(valueOrError)
        var defaultValue = txn.value
        if (value == null) {
            value = defaultValue
        }
        if (showNegativeOnly && value < 0) {
            // render values as positive
            value *= -1
            defaultValue *= -1
        } else if (showNegativeOnly || value < 0) {
            return ""
        }
        return if (value != defaultValue) {
            String.format(
                "%s (was %s)",
                formatCurrency(value),
                formatCurrency(defaultValue)
            )
        } else {
            formatCurrency(value)
        }
    }

    override fun toString(): String {
        return txn.description
    }

    fun getCellSource(col: Char): String? {
        return when (col) {
            'A' -> a
            'B' -> b
            'C' -> c
            'V' -> formula()
            else -> null
        }
    }

    companion object {
        fun toCents(value: Any?): Long? {
            return when (value) {
                is Double -> (100 * value).roundToLong()
                is Float -> (100 * value).roundToLong()
                is Long -> 100 * value
                is Int -> (100 * value).toLong()
                is BigInteger -> 100 * value.toLong()
                else -> null
            }
        }
    }
}